Programs/_freeze_module.c

```diff
@@ -98,39 +60,89 @@ main(int argc, char *argv[])
     if (PyStatus_Exception(status)) {
         Py_ExitStatusException(status);
     }
+}
+
+static const char *
+read_text(const char *inpath)
+{
+    FILE *infile = fopen(inpath, "rb");
+    if (infile == NULL) {
+        fprintf(stderr, "cannot open '%s' for reading\n", inpath);
+        return NULL;
+    }
+
+    struct _Py_stat_struct stat;
+    if (_Py_fstat_noraise(fileno(infile), &stat)) {
+        fprintf(stderr, "cannot fstat '%s'\n", inpath);
+        fclose(infile);
+        return NULL;
+    }
+    size_t text_size = (size_t)stat.st_size;
 
+    char *text = (char *) malloc(text_size + 1);
+    if (text == NULL) {
+        fprintf(stderr, "could not allocate %ld bytes\n", (long) text_size);
+        fclose(infile);
+        return NULL;
+    }
+    size_t n = fread(text, 1, text_size, infile);
+    fclose(infile);
+
+    if (n < text_size) {
+        fprintf(stderr, "read too short: got %ld instead of %ld bytes\n",
+                (long) n, (long) text_size);
+        free(text);
+        return NULL;
+    }
+
+    text[text_size] = '\0';
+    return (const char *)text;
+}
+
+static PyObject *
+compile_and_marshal(const char *name, const char *text)
+{
+    char buf[100];
     sprintf(buf, "<frozen %s>", name);
```

Comment:
The buffer overflow possibility here irks me -- shouldn't you at least verify that name is less than 90 chars?